pkg/platform/openstack/kuryr_bootstrap.go

```diff
@@ -542,6 +545,39 @@ func BootstrapKuryr(conf *operv1.NetworkSpec, kubeClient client.Client) (*bootst
 		}
 	}
 
+	octaviaVersion, err := getSavedAnnotation(kubeClient, names.KuryrOctaviaVersionAnnotation)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, errors.Wrap(err, "failed to get kuryr-config ConfigMap")
+	}
+
+	maxOctaviaVersion, err := getMaxOctaviaAPIVersion(lbClient)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get max octavia api version")
+	}
+
+	// In case the Kuryr config-map is annotated with an Octavia version different
+	// than the current Octavia version, and older than the version that multiple
+	// listeners becomes available and the Octavia provider is amphora, an Octavia
+	// upgrade happened and UDP listeners are now allowed to be created.
+	// By recreating the OpenShift DNS service a new load balancer amphora is in
+	// place with all required listeners.
+	log.Print("Checking Octavia upgrade happened")
+	if octaviaVersion != "" && octaviaProvider == "default" {
+		savedOctaviaVersion := semver.MustParse(octaviaVersion)
+		multipleListenersVersion := semver.MustParse(MinOctaviaVersionWithMultipleListeners)
+		if !savedOctaviaVersion.Equal(maxOctaviaVersion) && savedOctaviaVersion.LessThan(multipleListenersVersion) && octaviaMultipleListenersSupport {
+			dnsService := &v1.Service{
```

Comment:
Will this cause disruption to the cluster?

Reply:
It will cause a small downtime of few seconds on the DNS resolution, as a new load balancer with the double listeners is getting created.  After the load balancer is in place, the cluster should work normally.

Reply:
with amphora octavia driver I would say not just a few seconds, but around a minute. Note though this only happens upon OpenStack octavia upgrade between octavia version not supporting UDP to the one supporting it -- on the same port as the TCP one.

Reply:
That is... a bit concerning. What are the end-user expectations around this upgrade?

Reply:
The path for Octavia upgrade from version 13 to 16 is still in progress.
Octavia team suggestion for the problem being solved here is to make the operator open a support ticket requesting for the upgrade of a Load Balancer. However, we can't rely on that as our approach for enforcing DNS resolution over TCP would be gone as soon as the upgrade happens.

We expect to work on other approaches for this solution to make it more ideal in following releases.